dependencies/aiven-kafka-auth/src/main/java/io/aiven/kafka/auth/AivenSaslScramServerCallbackHandler.java

```diff
@@ -0,0 +1,104 @@
+package io.aiven.kafka.auth;
+
+import java.io.File;
+import java.io.FileReader;
+import java.io.IOException;
+import java.security.NoSuchAlgorithmException;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import javax.security.auth.callback.Callback;
+import javax.security.auth.callback.NameCallback;
+import javax.security.auth.callback.UnsupportedCallbackException;
+import javax.security.auth.login.AppConfigurationEntry;
+import org.apache.kafka.common.KafkaException;
+import org.apache.kafka.common.security.JaasContext;
+import org.apache.kafka.common.security.auth.AuthenticateCallbackHandler;
+import org.apache.kafka.common.security.scram.ScramCredential;
+import org.apache.kafka.common.security.scram.ScramCredentialCallback;
+import org.apache.kafka.common.security.scram.ScramLoginModule;
+import org.apache.kafka.common.security.scram.internals.ScramFormatter;
+import org.apache.kafka.common.security.scram.internals.ScramMechanism;
+import org.json.simple.JSONArray;
+import org.json.simple.JSONObject;
+import org.json.simple.parser.JSONParser;
+import org.json.simple.parser.ParseException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class AivenSaslScramServerCallbackHandler implements AuthenticateCallbackHandler {
+  private static final Logger logger =
+      LoggerFactory.getLogger(AivenSaslScramServerCallbackHandler.class);
+  private String configFileLocation;
+  private ScramFormatter formatter;
+
+  @Override
+  public void configure(Map<String, ?> configs, String mechanism,
+          List<AppConfigurationEntry> jaasConfigEntries) {
+    configFileLocation = JaasContext.configEntryOption(
+            jaasConfigEntries, "users.config", ScramLoginModule.class.getName());
+    logger.info("Using configuration file {}", configFileLocation);
+    try {
+      formatter = new ScramFormatter(ScramMechanism.forMechanismName(mechanism));
+    } catch (NoSuchAlgorithmException e) {
+      logger.error("Error configuring SASL/SCRAM callback, unrecognized mechanism {}", mechanism);
+    }
+  }
+
+  @Override
+  public void handle(Callback[] callbacks) throws IOException, UnsupportedCallbackException {
+    String username = null;
+    for (Callback callback : callbacks) {
+      if (callback instanceof NameCallback) {
+        NameCallback nameCallback = (NameCallback) callback;
+        username = nameCallback.getDefaultName();
+      } else if (callback instanceof ScramCredentialCallback) {
+        ScramCredentialCallback credCallback = (ScramCredentialCallback) callback;
+        ScramCredential creds = getScramCreds(username);
+        credCallback.scramCredential(creds);
+      } else {
+        throw new UnsupportedCallbackException(callback);
+      }
+    }
+  }
+
+  /** Generate SCRAM creds for given username. */
+  public ScramCredential getScramCreds(String username) {
+    File configFile = new File(configFileLocation);
```

Comment:
Reading and parsing the file every time is costly. We should do similar caching we do in `AivenKafkaPrincipalBuilder`. (Read full file contents to memory, every 10 seconds (or some other configurable value) check file modified timestamp, if changed read in new contents.) I see this same logic is currently in `AivenSaslPlainServerCallbackHandler` so not a regression from that point-of-view but would be good to have both of them do caching. The OVO experience showed inefficient logic in Kafka auth can be costly.

Reply:
Let's leave this one for later.

Reply:
For completeness sake, these comments were from the PR in the original repo:

me: I thought a new instance of principal mapper was created for each channel / connection, and didn't realize such a trivial caching was possible. I'l address this.

ivanyu: From the channel's code it is really so. It just isn't documented anyhow. That's why in AivenSaslPlainServerCallbackHandler we played safe. I asked this in the mailing list, but no-one replied.